File: runtime/typescript/packages/core/src/core/memory.ts
import { MemoryEntry, MemoryStore } from "../model/index.js";

export interface ScoredMemory {
  entry: MemoryEntry;
  score: number;
  keywordMatches: number;
}

export interface MemoryPort {
  load(): MemoryStore;
  save(store: MemoryStore): void;
}

export function addMemory(store: MemoryStore, entry: MemoryEntry): void {
  store.entries.push(entry);
}

export function remember(
  store: MemoryStore,
  entry: MemoryEntry,
  maxEntries = 0,
): void {
  if (entry.category === "core") {
    store.entries = store.entries.filter(
      (existing) =>
        existing.category !== "core" || !tagsEqual(existing.tags, entry.tags),
    );
  }
  store.entries.push(entry);
  evictToCap(store, maxEntries);
}

export function evictToCap(store: MemoryStore, maxEntries: number): number {
  if (maxEntries <= 0) {
    return 0;
  }
  let evicted = 0;
  while (store.entries.length > maxEntries) {
    const archivalIndex = store.entries.findIndex(
      (entry) => entry.category === "archival",
    );
    store.entries.splice(archivalIndex >= 0 ? archivalIndex : 0, 1);
    evicted += 1;
  }
  return evicted;
}

export function updateMemory(
  store: MemoryStore,
  index: number,
  entry: MemoryEntry,
): void {
  requireIndex(store, index);
  store.entries[index] = entry;
}

export function updateContent(
  store: MemoryStore,
  index: number,
  content: string,
): void {
  requireIndex(store, index);
  store.entries[index]!.content = content;
}

export function removeMemory(store: MemoryStore, index: number): MemoryEntry {
  requireIndex(store, index);
  return store.entries.splice(index, 1)[0]!;
}

export function clearMemory(store: MemoryStore, category?: string): number {
  const before = store.entries.length;
  if (category === undefined || category === null || category === "") {
    store.entries = [];
  } else {
    store.entries = store.entries.filter(
      (entry) => entry.category !== category,
    );
  }
  return before - store.entries.length;
}

export function coreMemories(store: MemoryStore): MemoryEntry[] {
  return store.entries.filter((entry) => entry.category === "core");
}

export function recall(
  store: MemoryStore,
  query: string,
  limit = 0,
): ScoredMemory[] {
  const tokens = queryTokens(query);
  const hasQuery = tokens.length > 0;
  const scored: ScoredMemory[] = [];

  for (const entry of store.entries) {
    const { score, keywordMatches } = scoreEntry(entry, tokens);
    if (hasQuery && keywordMatches === 0) {
      continue;
    }
    scored.push({ entry, score, keywordMatches });
  }

  scored.sort((left, right) => right.score - left.score);
  return limit > 0 ? scored.slice(0, limit) : scored;
}

export function formatForSystemPrompt(store: MemoryStore): string {
  const core = coreMemories(store);
  if (core.length === 0) {
    return "";
  }
  return `## Memory\n${core.map((entry) => `- ${entry.content}\n`).join("")}`;
}

export function formatRecallResults(results: ScoredMemory[]): string {
  if (results.length === 0) {
    return "";
  }
  return results
    .map((result, index) => {
      let line = `${index + 1}. [${result.entry.category}] ${result.entry.content}\n`;
      if (result.entry.tags && result.entry.tags.length > 0) {
        line += `   tags: ${result.entry.tags.join(", ")}\n`;
      }
      return line;
    })
    .join("");
}

function requireIndex(store: MemoryStore, index: number): void {
  if (index < 0 || index >= store.entries.length) {
    throw new RangeError(
      `memory index ${index} out of bounds (len ${store.entries.length})`,
    );
  }
}

function tagsEqual(left?: string[], right?: string[]): boolean {
  const a = left ?? [];
  const b = right ?? [];
  return a.length === b.length && a.every((value, index) => value === b[index]);
}

function queryTokens(query: string): string[] {
  const tokens: string[] = [];
  for (const raw of (query ?? "").split(/\s+/u)) {
    const trimmed = trimNonAlphanumeric(raw).toLowerCase();
    if (trimmed.length > 0 && !tokens.includes(trimmed)) {
      tokens.push(trimmed);
    }
  }
  return tokens;
}

function trimNonAlphanumeric(value: string): string {
  return value.replace(/^[^\p{L}\p{N}]+|[^\p{L}\p{N}]+$/gu, "");
}

function scoreEntry(
  entry: MemoryEntry,
  tokens: string[],
): { score: number; keywordMatches: number } {
  if (tokens.length === 0) {
    return { score: 0, keywordMatches: 0 };
  }
  const content = (entry.content ?? "").toLowerCase();
  const tags = (entry.tags ?? []).map((tag) => tag.toLowerCase());
  let score = 0;
  let keywordMatches = 0;

  for (const token of tokens) {
    const inContent = content.includes(token);
    const inTags = tags.some((tag) => tag.includes(token));
    if (inContent || inTags) {
      keywordMatches += 1;
    }
    if (inContent) {
      score += 2;
    }
    if (inTags) {
      score += 3;
    }
  }
  if (score > 0 && entry.category === "core") {
    score += 1;
  }
  return { score, keywordMatches };
}

File: runtime/swift/prompty-model/Sources/PromptyModel/memory.swift
import Foundation

public struct ScoredMemory {
  public var entry: MemoryEntry
  public var score: Double
  public var keywordMatches: Int

  public init(entry: MemoryEntry, score: Double, keywordMatches: Int) {
    self.entry = entry
    self.score = score
    self.keywordMatches = keywordMatches
  }
}

public protocol MemoryPort {
  func load() -> MemoryStore
  func save(_ store: MemoryStore)
}

public func addMemory(_ store: inout MemoryStore, _ entry: MemoryEntry) {
  store.entries.append(entry)
}

public func remember(_ store: inout MemoryStore, _ entry: MemoryEntry, maxEntries: Int = 0) {
  if entry.category == .core {
    store.entries.removeAll { existing in
      existing.category == .core && tagsEqual(existing.tags, entry.tags)
    }
  }
  store.entries.append(entry)
  _ = evictToCap(&store, maxEntries: maxEntries)
}

public func evictToCap(_ store: inout MemoryStore, maxEntries: Int) -> Int {
  if maxEntries <= 0 { return 0 }
  var evicted = 0
  while store.entries.count > maxEntries {
    let victim = store.entries.firstIndex { $0.category == .archival } ?? 0
    store.entries.remove(at: victim)
    evicted += 1
  }
  return evicted
}

public func updateMemory(_ store: inout MemoryStore, index: Int, entry: MemoryEntry) throws {
  try requireIndex(store, index)
  store.entries[index] = entry
}

public func updateContent(_ store: inout MemoryStore, index: Int, content: String) throws {
  try requireIndex(store, index)
  store.entries[index].content = content
}

public func removeMemory(_ store: inout MemoryStore, index: Int) throws -> MemoryEntry {
  try requireIndex(store, index)
  return store.entries.remove(at: index)
}

public func clearMemory(_ store: inout MemoryStore, category: MemoryCategory? = nil) -> Int {
  let before = store.entries.count
  if let category {
    store.entries.removeAll { $0.category == category }
  } else {
    store.entries.removeAll()
  }
  return before - store.entries.count
}

public func coreMemories(_ store: MemoryStore) -> [MemoryEntry] {
  store.entries.filter { $0.category == .core }
}

public func recall(_ store: MemoryStore, query: String, limit: Int = 0) -> [ScoredMemory] {
  let tokens = queryTokens(query)
  let hasQuery = !tokens.isEmpty
  var scored: [ScoredMemory] = []

  for entry in store.entries {
    let result = scoreEntry(entry, tokens: tokens)
    if hasQuery && result.keywordMatches == 0 { continue }
    scored.append(ScoredMemory(entry: entry, score: result.score, keywordMatches: result.keywordMatches))
  }

  scored = scored.enumerated()
    .sorted { left, right in
      if left.element.score == right.element.score { return left.offset < right.offset }
      return left.element.score > right.element.score
    }
    .map(\.element)
  if limit > 0 && scored.count > limit {
    return Array(scored.prefix(limit))
  }
  return scored
}

public func formatForSystemPrompt(_ store: MemoryStore) -> String {
  let core = coreMemories(store)
  if core.isEmpty { return "" }
  return "## Memory\n" + core.map { "- \($0.content)\n" }.joined()
}

public func formatRecallResults(_ results: [ScoredMemory]) -> String {
  if results.isEmpty { return "" }
  return results.enumerated().map { index, result in
    var line = "\(index + 1). [\(result.entry.category.rawValue)] \(result.entry.content)\n"
    if let tags = result.entry.tags, !tags.isEmpty {
      line += "   tags: \(tags.joined(separator: ", "))\n"
    }
    return line
  }.joined()
}

private func requireIndex(_ store: MemoryStore, _ index: Int) throws {
  if index < 0 || index >= store.entries.count {
    throw MemoryError.indexOutOfRange("memory index \(index) out of bounds (len \(store.entries.count))")
  }
}

public enum MemoryError: Error {
  case indexOutOfRange(String)
}

private func tagsEqual(_ left: [String]?, _ right: [String]?) -> Bool {
  (left ?? []) == (right ?? [])
}

private func queryTokens(_ query: String) -> [String] {
  var tokens: [String] = []
  for raw in query.split(whereSeparator: { $0.isWhitespace }) {
    let token = trimNonAlphanumeric(String(raw)).lowercased()
    if !token.isEmpty && !tokens.contains(token) {
      tokens.append(token)
    }
  }
  return tokens
}

private func trimNonAlphanumeric(_ value: String) -> String {
  var scalars = value.unicodeScalars[...]
  while let first = scalars.first, !CharacterSet.alphanumerics.contains(first) {
    scalars.removeFirst()
  }
  while let last = scalars.last, !CharacterSet.alphanumerics.contains(last) {
    scalars.removeLast()
  }
  return String(String.UnicodeScalarView(scalars))
}

private func scoreEntry(_ entry: MemoryEntry, tokens: [String]) -> (score: Double, keywordMatches: Int) {
  if tokens.isEmpty { return (0, 0) }
  let content = entry.content.lowercased()
  let tags = (entry.tags ?? []).map { $0.lowercased() }
  var score = 0.0
  var keywordMatches = 0

  for token in tokens {
    let inContent = content.contains(token)
    let inTags = tags.contains { $0.contains(token) }
    if inContent || inTags {
      keywordMatches += 1
    }
    if inContent {
      score += 2
    }
    if inTags {
      score += 3
    }
  }
  if score > 0 && entry.category == .core {
    score += 1
  }
  return (score, keywordMatches)
}
